src/hash_builder/input.rs

```diff
@@ -0,0 +1,107 @@
+use alloc::vec::Vec;
+use alloy_primitives::{hex, B256};
+use core::fmt;
+
+/// The input of the hash builder.
+///
+/// Stores [`HashBuilderInputRef`] efficiently by reusing resources.
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(derive_arbitrary::Arbitrary, proptest_derive::Arbitrary))]
+pub struct HashBuilderInput {
+    /// Stores the bytes of either the leaf node value or the hash of adjacent nodes.
+    buf: Vec<u8>,
+    /// The kind of the current hash builder input.
+    kind: HashBuilderInputKind,
+}
+
+impl Default for HashBuilderInput {
+    fn default() -> Self {
+        Self { buf: Vec::with_capacity(128), kind: HashBuilderInputKind::default() }
+    }
+}
+
+impl fmt::Debug for HashBuilderInput {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.as_ref().fmt(f)
+    }
+}
+
+impl HashBuilderInput {
+    /// Returns the input as a reference.
+    #[inline]
+    pub fn as_ref(&self) -> HashBuilderInputRef<'_> {
+        match self.kind {
+            HashBuilderInputKind::Bytes => HashBuilderInputRef::Bytes(&self.buf),
+            HashBuilderInputKind::Hash => {
+                debug_assert_eq!(self.buf.len(), 32);
+                HashBuilderInputRef::Hash(unsafe { self.buf[..].try_into().unwrap_unchecked() })
+            }
+        }
+    }
+
+    /// Sets the input from the given bytes.
+    #[inline]
+    pub fn set_from_ref(&mut self, input: HashBuilderInputRef<'_>) {
+        self.buf.clear();
+        self.buf.extend_from_slice(input.as_slice());
+        self.kind = input.kind();
+    }
+
+    /// Clears the input.
+    #[inline]
+    pub fn clear(&mut self) {
+        self.buf.clear();
+        self.kind = HashBuilderInputKind::default();
+    }
+}
+
+/// The kind of the current hash builder input.
+#[doc(hidden)] // Not public API.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "arbitrary", derive(derive_arbitrary::Arbitrary, proptest_derive::Arbitrary))]
+#[allow(unreachable_pub)]
+pub enum HashBuilderInputKind {
```

Comment:
why the input rename? nodes have values, not inputs

Reply:
was thinking to include key or something i forgot, no reason now